README.md
# Introduction

This project demonstrates the E2E tests

# How to use

This e2e test is done using CodeceptJS https://codecept.io/

### Tech

e2e test uses a number of open source projects to work properly:

* https://nodejs.org/en/ - evented I/O for the backend
* https://codecept.io/ - CodeceptJS

Additionally, the tests are executed via [Browserstack](https://www.browserstack.com) and results are updated accordingly with this helper [codeceptjs](https://www.npmjs.com/package/codeceptjs-bshelper)

### Installation

e2e test requires [Node.js](https://nodejs.org/) v8+ to run.

Install the dependencies and devDependencies.

```sh
$ cd aklamio
$ npm install
```

### How to trigger UI tests

For all possible combinations of browsers and OSes, please refer to `codecept.conf.js`

- To run the tests with Chrome 74 on Windows 10

```sh
BROWSER=chrome_74_win_10 npx codeceptjs run
```

- To run the tests with FF 66 on Windows 10

```sh
BROWSER=firefox_66_win_10 npx codeceptjs run
```

- To run the tests with Safari 12 on MacOS Mojave

```sh
BROWSER=safari_12_mojave npx codeceptjs run
```

- To run the tests with IE 11 on Windows 10

```sh
BROWSER=ie_11_win_10 npx codeceptjs run
```

### Travis CI

This project is also configured to be triggered in Travis CI/ Gitlab CI whenever we push a new commit. Tests will be run against 4 browsers: IE11, Safari, Firefox and Chrome.

What you can see after the TravisCI is done
![Travis-CI](https://d1sz9tkli0lfjq.cloudfront.net/items/0q432S3R0x0I2e182J47/Image%202019-05-07%20at%203.56.45%20PM.png?v=8a850bbd)

To see more details, clicking on any job, for instance:
```sh
84.91s$ BROWSER=${BROWSER} npx codeceptjs run
creating output directory: /home/travis/build/PeterNgTr/aklamio/output
CodeceptJS v2.1.1
Using test root "/home/travis/build/PeterNgTr/aklamio"
Login --
✔ Login with invalid credentials in 13851ms
Test has Passed
Test finished. Link to job: https://automate.browserstack.com/builds/2815b335c0674ab0b5eb0b2bd6f53ead2f92f953/sessions/8526a2358f5d1595723a49c083b03a0fa8c2945a?auth_token=4660aacfcf0a9ae19b44c8953eeb5d10c578ddf8c0620de4f11e8835d57c28df
Referral --
✔ Refer product 1 successfully in 16811ms
Test has Passed
Test finished. Link to job: https://automate.browserstack.com/builds/2815b335c0674ab0b5eb0b2bd6f53ead2f92f953/sessions/02cc287a3e5d00ef6a06284f1d27b1d884c05796?auth_token=e7ace75d3be85aa309eb4c5b82beb3a9351e41818e78f6d98957de805b066c4f
✔ Refer product 1 with invalid email in 16273ms
Test has Passed
Test finished. Link to job: https://automate.browserstack.com/builds/2815b335c0674ab0b5eb0b2bd6f53ead2f92f953/sessions/0cc3cca50c811a6a0a6df3aabda28281d45b51c5?auth_token=8cc15db8547c34d92028c8561e4f96f011f76a133deb1c5d9197bce7c3c8dd2a
  OK  | 3 passed   // 1m
The command "BROWSER=${BROWSER} npx codeceptjs run" exited with 0.
```

### Test Results in BrowserStack
Our tests are executed against BrowserStack, hence you will see the results from there, in case of failed tests, you can see detail logs.
![BS-Automate](https://d1sz9tkli0lfjq.cloudfront.net/items/3y380t342O1x1i1H0a1Y/Image%202019-05-07%20at%204.01.05%20PM.png?v=b4a92e01)

Moreover you can find the public link of run tests that are exposed in TravisCI logs and this eases us much in case of sharing to others.

```sh
✔ Refer product 1 with invalid email in 16273ms
Test has Passed
Test finished. Link to job: https://automate.browserstack.com/builds/2815b335c0674ab0b5eb0b2bd6f53ead2f92f953/sessions/0cc3cca50c811a6a0a6df3aabda28281d45b51c5?auth_token=8cc15db8547c34d92028c8561e4f96f011f76a133deb1c5d9197bce7c3c8dd2a
```







HISTORY.md

### Initialization

Once receiving this task, I feel much excited to work with, to be honest, any challenge I always love to deal with cause throughout of it I could learn more and more.

Well, firstly, of course I need to read throughout the task to learn what I should do. Okay then I will need to create an automation tests that cover 3 test cases and a bonus point to integrate my tests with a CI/CD.

Let deep dive in the task, as a first step, after having clear about the task, I started to do the manual tests with all the test cases that I need to implement the automation flows for it. After having the manual flows now it's time to start the implementation.

### Choosing framework

Having trying many frameworks, now, I found out a favorite framework of mine, it's called [CodeceptJS](https://codecept.io/). The reason I love this cause CodeceptJS is multi-backend testing framework. It can execute tests using different libraries like `webdriverio`, `Puppeteer`, `Protractor`, etc. or literally meaning `helpers` you can switch the helper as your choice. Also it supports many things more which you can find in their webpage. I'm also an active contributor on that open source project, I implemented/fixing many cool things there. One of this is [BrowserStack Helper](https://www.npmjs.com/package/codeceptjs-bshelper) which is used in this task to update `Test Names`, `Test Results` after execution.

### Organization

Well, I decided to structure the tests as page objects models, having said that, I mean, we should keep our tests and element locators separately, this will keep code clean and easy to understand and maintain,  any change in UI can easily be implemented, updated and maintained into the Page Objects and Classes. That you will see in project setup, a folder called `pages` where page objects sit and `test scenarios` will sit in `test` folder. 

Each page object will keep `element locators` and `functions` that can be called and reused anywhere it is needs.

### Implementation
Now it's time to begin the journey, capturing locators is not an easy game, you need to think of how to make it efficiently. Everything went well. I started the run the first test. Oh no, it's failing. Haha, I didn't expect everything passed after the first try. That's weird, the locator is proper, why the console logs told that element is not located. I tried the same locator with `chrome dev tools`, it's working. All right, after some times I found out, we had an `iframe`. Aha,  we need to switch to an `iframe` before locating elements. Phew, done. My test case was now passed. Amazing! I finished the rest of test cases.

### Cross-browsers testing
The beauty of web testing, haha, I would say, is cross-browsers testing, if your tests are passed on Chrome, it doesn't mean it will pass on other browsers. Now [BrowserStack](https://www.browserstack.com) enters the story. The good test would be covered as many browsers as possible. Then I updated the codeceptjs config to make the tests can be executed with many combinations of browsers in BrowserStack. My tests are working perfectly with
- BROWSER=ie_11_win_10
- BROWSER=chrome_74_win_10
- BROWSER=firefox_66_win_10
- BROWSER=safari_12_mojave

### CI/CD
That is even though a bonus point, but to me, I'm always curious how my tests are running in CI/CD. I chose `TravisCI` to start with, having some experiences with other CI/CD like `Jenkins`, `TeamCity`, `Gitlab` but I would say `TravisCI` easy to play with compared to others as an open source project, integration is quite simple. You can refer to this for more details how the test are running [https://travis-ci.org/PeterNgTr/aklamio](https://travis-ci.org/PeterNgTr/aklamio)

*Note*: To keep the free version `TravisCI` works with this, I will leave the project as `public`. 

### Test Results

An important thing after the execution is test results. You can either check the test results from `TravisCI` or `BrowserStack`

Our tests are executed against BrowserStack, hence you will see the results from there, in case of failed tests, you can see detail logs.

![BS-Automate](https://d1sz9tkli0lfjq.cloudfront.net/items/3y380t342O1x1i1H0a1Y/Image%202019-05-07%20at%204.01.05%20PM.png?v=b4a92e01)

 
Moreover you can find the public link of run tests that are exposed in TravisCI logs and this eases us much in case of sharing to others. Thanks to my custom helper [BrowserStack Helper](https://www.npmjs.com/package/codeceptjs-bshelper) 

 
```sh

✔ Refer product 1 with invalid email in 16273ms

Test has Passed

Test finished. Link to job: https://automate.browserstack.com/builds/2815b335c0674ab0b5eb0b2bd6f53ead2f92f953/sessions/0cc3cca50c811a6a0a6df3aabda28281d45b51c5?auth_token=8cc15db8547c34d92028c8561e4f96f011f76a133deb1c5d9197bce7c3c8dd2a

```

data_test.js
module.exports = {
    email: {
        valid: 'peter.nguyentr@gmail.com',
        invalid: 'cbbh@cnj.c'
    },
    credentials: {
        invalid: {
            email: 'peter.nguyentr@gmail.com',
            password: '123456'
        }
    }
}

test/login_test.js
const { I, integrationPage, data } = inject();

Feature('Login');

Before(() => {
    I.amOnPage(`/integration/${process.env.PORTAL_ID}`);
})

Scenario('Login with invalid credentials', () => {
    const { email, password } = data.credentials.invalid 
    integrationPage.login({ email, password });
    integrationPage.seeError();
});



test/referral_test.js
const { I, integrationPage, referralModalBox, data } = inject();

Feature('Referral');

Before(() => {
    I.amOnPage(`/integration/${process.env.PORTAL_ID}`);
})

Scenario('Refer product 1 successfully', () => {
    integrationPage.referNow();
    referralModalBox.referProduct(data.email.valid);
    referralModalBox.seeSharingLink();
    referralModalBox.seeSocialSharingOptions();
});

Scenario('Refer product 1 with invalid email', () => {
    integrationPage.referNow();
    referralModalBox.referProduct(data.email.invalid);
    referralModalBox.seeError();
});


pages/integrationPage.js
const I = actor();

export default integrationPage = {

  referNowBtn: 'div[data-form="register"] button',
  loginLnk: 'a[href="#"]',
  fields: {
    email: 'input[data-field="email"]',
    password: 'input[data-field="password"]'
  },
  loginBtn: 'button[type="submit"]',
  errorMsg: 'div[class="notification error"]',

  referNow() {
    I.wait(3);
    within({ frame: `${process.env.IFRAME_ID}` }, () => {
      I.click(this.referNowBtn);
    });
  },

  login(data) {
    I.wait(3);
    within({ frame: `${process.env.IFRAME_ID}` }, () => {
      I.click(this.loginLnk);
      I.fillField(this.fields.email, data.email);
      I.fillField(this.fields.password, data.password);
      I.click(this.loginBtn);
    });
  },

  seeError() {
    within({ frame: `${process.env.IFRAME_ID}` }, () => {
      I.seeElement(this.errorMsg);
    });
  }
}


pages/referralModalbox.js
const I = actor();

export default referralModalBox = {

  giveawayBtn: 'button[data-aklamio-recommend-rr-follower]',
  emailField: 'form[data-form="find_or_register"] input[data-field="email"]',
  continueBtn: 'form[data-form="find_or_register"] button[type="submit"]',
  sharingLink: 'input[data-aklamio-recommend-shortlink]',
  socialSharingOptions: 'ul[class="referral-modalbox-sharing"]',
  errorMsg: 'div[class="notification error"]',

  referProduct(emailAddress, type = 'giveaway') {
    within({ frame: `${process.env.IFRAME_ID}` }, () => {
      switch (type) {
        case 'giveaway':
          I.click(this.giveawayBtn);
          break;
      }

      I.fillField(this.emailField, emailAddress);
      I.click(this.continueBtn);
      I.wait(2);
    });
  },

  seeSharingLink() {
    within({ frame: `${process.env.IFRAME_ID}` }, () => {
      I.seeElement(this.sharingLink);
    });
  },

  seeSocialSharingOptions() {
    within({ frame: `${process.env.IFRAME_ID}` }, () => {
      I.seeElement(this.socialSharingOptions);
    });
  },

  seeError() {
    within({ frame: `${process.env.IFRAME_ID}` }, () => {
      I.seeElement(this.errorMsg);
    });
  }
}


codecept.conf.js
require('import-export');

process.env.IFRAME_ID = `[src="https://sonata.aklamio.com/portals/${process.env.PORTAL_ID}/index.html#route=index"]`;

const _user = process.env.BROWSERSTACK_USERNAME;
const _key = process.env.BROWSERSTACK_ACCESS_KEY;

const commonBrowserstackCapabilities = {
  'browserstack.debug': 'true',
  resolution: '1920x1080',
  project: 'Aklamio',
  build: 'Aklamio Master',
  'browserstack.timezone': 'CEST'
};

const BROWSERS = {
  chrome: [74],
  firefox: [66],
  safari: [12],
  edge: [14, 15, 16],
  ie: [11]
};

const OPERATING_SYSTEMS = {
  WINDOWS: [{ version: '7', tag: 'win_7' }, { version: '8', tag: 'win_8' }, { version: '8.1', tag: 'win_8.1' }, { version: '10', tag: 'win_10' }],
  'OS X': [{ version: 'Mojave', tag: 'mojave' }, { version: 'High Sierra', tag: 'sierra' }, { version: 'El Capitan', tag: 'capitan' }]

};

const browserMatrix = {
  // mobile web capabilities
  chrome_mobile: {
      desiredCapabilities: {
          os: 'OS X',
          os_version: 'Mojave',
          browser: 'Chrome',
          browser_version: '72',
          browserName: 'chrome_mobile',
          realMobile: 'true',
          chromeOptions: {
              mobileEmulation: { deviceName: 'Pixel 2' },
              args: ['window-size=400,800']
          }
      }
  }
};

Object.keys(BROWSERS).forEach(browserName => {
  BROWSERS[browserName].forEach(version => {
      for (const os in OPERATING_SYSTEMS) {
          OPERATING_SYSTEMS[os].forEach(osVersion => {
              browserMatrix[`${browserName}_${version}_${osVersion.tag}`] = {
                  desiredCapabilities: {
                      browser: browserName,
                      browserName,
                      browser_version: version,
                      os,
                      os_version: osVersion.version
                  }
              };
          });
      }
  });
});

Object.keys(browserMatrix).forEach(key => {
  Object.assign(browserMatrix[key].desiredCapabilities, commonBrowserstackCapabilities);
});

exports.config = {
  tests: './test/*.js',
  output: './output',
  helpers: {
    WebDriver: {
      url: 'https://sonata.aklamio.info',
      windowSize: '1980x1080',
      smartWait: 5000,
      waitForTimeout: 5000,
      browser: 'chrome',
      user: _user,
      key: _key,
      host: 'hub.browserstack.com',
      port: 443,
      path: '/wd/hub',
      desiredCapabilities: browserMatrix[process.env.BROWSER].desiredCapabilities
    },
    BrowserstackHelper: {
      require: "codeceptjs-bshelper",
      user: _user,
      key: _key
    },
    REST: {},
  },
  include: {
    I: './steps_file',
    integrationPage: './pages/integration',
    referralModalBox: './pages/referralModal',
    data: './data_test',
  },
  mocha: {},
  name: 'aklamio',
  plugins: {
    wdio: {
      enabled: false,
      services: ['selenium-standalone']
    }
  },
  multiple: {
    parallel: {
        chunks: (files) => {
            return files.map(file => [file]);
        }
    }
},
}